llvm/lib/Transforms/Scalar/ConstraintElimination.cpp

```diff
@@ -900,10 +905,88 @@ static void dumpConstraint(ArrayRef<int64_t> C,
 }
 #endif
 
+static Value *getStartValueFromAddRec(PHINode &PN, Loop &L,
+                                      ScalarEvolution &SE) {
+  auto *AR = dyn_cast_or_null<SCEVAddRecExpr>(SE.getSCEV(&PN));
```

Comment:
I think this is only called in contexts where this cast is guaranteed to succeed. So I'd probably just pass the existing `AR` as a parameter (or at least use an unchecked `cast<>` instead of a `dyn_cast_or_null<>`).

Reply:
Fixed